
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://combine-lab.github.io/QCatch/">
      
      
      
        <link rel="next" href="demo/demo/">
      
      
        
      
      
      <link rel="icon" href="assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.2">
    
    
      
        <title>Documentation for QCatch</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.484c7ddc.min.css">
      
        
        <link rel="stylesheet" href="assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#qcatch" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="." title="Documentation for QCatch" class="md-header__button md-logo" aria-label="Documentation for QCatch" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Documentation for QCatch
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Home
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/COMBINE-lab/QCatch" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="." title="Documentation for QCatch" class="md-nav__button md-logo" aria-label="Documentation for QCatch" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    Documentation for QCatch
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/COMBINE-lab/QCatch" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="." class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#installation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Installation
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Installation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-bioconda" class="md-nav__link">
    <span class="md-ellipsis">
      
        1. Bioconda
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-pypi" class="md-nav__link">
    <span class="md-ellipsis">
      
        2. PyPI
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#basic-usage" class="md-nav__link">
    <span class="md-ellipsis">
      
        Basic Usage
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tutorial-run-qcatch-on-example-data" class="md-nav__link">
    <span class="md-ellipsis">
      
        Tutorial: Run QCatch on Example data
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Tutorial: Run QCatch on Example data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#step-1-download-dataset" class="md-nav__link">
    <span class="md-ellipsis">
      
        Step 1 â€” Download Dataset
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-2-run-the-qcatch" class="md-nav__link">
    <span class="md-ellipsis">
      
        Step 2 - Run the qcatch
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tips" class="md-nav__link">
    <span class="md-ellipsis">
      
        Tips
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Tips">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-input-path" class="md-nav__link">
    <span class="md-ellipsis">
      
        1- Input path:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-output-path" class="md-nav__link">
    <span class="md-ellipsis">
      
        2- Output path:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-chemistry" class="md-nav__link">
    <span class="md-ellipsis">
      
        3- Chemistry:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-gene-gene-mapping-file" class="md-nav__link">
    <span class="md-ellipsis">
      
        4- Gene gene mapping file:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-save-filtered-h5ad-file" class="md-nav__link">
    <span class="md-ellipsis">
      
        5- Save filtered h5ad file:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#6-specify-your-desired-cell-list" class="md-nav__link">
    <span class="md-ellipsis">
      
        6- Specify your desired cell list:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#7-skip-clustering-plots" class="md-nav__link">
    <span class="md-ellipsis">
      
        7- Skip clustering plots:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#8-export-the-summary-metrics" class="md-nav__link">
    <span class="md-ellipsis">
      
        8- Export the summary metrics
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#9-visualize-doublets" class="md-nav__link">
    <span class="md-ellipsis">
      
        9- Visualize doublets
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#10-debug-level-message" class="md-nav__link">
    <span class="md-ellipsis">
      
        10- Debug-level message
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#11-re-run-qcatch-on-modified-h5ad-file" class="md-nav__link">
    <span class="md-ellipsis">
      
        11- Re-run QCatch on modified h5ad file
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#example-directory-structures" class="md-nav__link">
    <span class="md-ellipsis">
      
        Example directory structures:
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#command-line-arguments" class="md-nav__link">
    <span class="md-ellipsis">
      
        Command-Line Arguments
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Demos
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Demos
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="demo/demo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="demo/demo.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Demo(1k PBMC)
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="demo/demo_low_depth.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Demo(1k PBMC-Low depth)
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="changelog/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Changelog
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="faq/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    FAQ
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#installation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Installation
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Installation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-bioconda" class="md-nav__link">
    <span class="md-ellipsis">
      
        1. Bioconda
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-pypi" class="md-nav__link">
    <span class="md-ellipsis">
      
        2. PyPI
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#basic-usage" class="md-nav__link">
    <span class="md-ellipsis">
      
        Basic Usage
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tutorial-run-qcatch-on-example-data" class="md-nav__link">
    <span class="md-ellipsis">
      
        Tutorial: Run QCatch on Example data
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Tutorial: Run QCatch on Example data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#step-1-download-dataset" class="md-nav__link">
    <span class="md-ellipsis">
      
        Step 1 â€” Download Dataset
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-2-run-the-qcatch" class="md-nav__link">
    <span class="md-ellipsis">
      
        Step 2 - Run the qcatch
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tips" class="md-nav__link">
    <span class="md-ellipsis">
      
        Tips
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Tips">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-input-path" class="md-nav__link">
    <span class="md-ellipsis">
      
        1- Input path:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-output-path" class="md-nav__link">
    <span class="md-ellipsis">
      
        2- Output path:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-chemistry" class="md-nav__link">
    <span class="md-ellipsis">
      
        3- Chemistry:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-gene-gene-mapping-file" class="md-nav__link">
    <span class="md-ellipsis">
      
        4- Gene gene mapping file:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-save-filtered-h5ad-file" class="md-nav__link">
    <span class="md-ellipsis">
      
        5- Save filtered h5ad file:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#6-specify-your-desired-cell-list" class="md-nav__link">
    <span class="md-ellipsis">
      
        6- Specify your desired cell list:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#7-skip-clustering-plots" class="md-nav__link">
    <span class="md-ellipsis">
      
        7- Skip clustering plots:
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#8-export-the-summary-metrics" class="md-nav__link">
    <span class="md-ellipsis">
      
        8- Export the summary metrics
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#9-visualize-doublets" class="md-nav__link">
    <span class="md-ellipsis">
      
        9- Visualize doublets
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#10-debug-level-message" class="md-nav__link">
    <span class="md-ellipsis">
      
        10- Debug-level message
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#11-re-run-qcatch-on-modified-h5ad-file" class="md-nav__link">
    <span class="md-ellipsis">
      
        11- Re-run QCatch on modified h5ad file
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#example-directory-structures" class="md-nav__link">
    <span class="md-ellipsis">
      
        Example directory structures:
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#command-line-arguments" class="md-nav__link">
    <span class="md-ellipsis">
      
        Command-Line Arguments
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="qcatch">QCatch</h1>
<p><a href="https://pypi.org/project/qcatch/"><img alt="PyPI version" src="https://img.shields.io/pypi/v/qcatch?cacheSeconds=3600&amp;logo=pypi&amp;label=PyPI" /></a>
<a href="https://github.com/COMBINE-lab/QCatch/actions/workflows/test.yaml"><img alt="Tests" src="https://img.shields.io/github/actions/workflow/status/COMBINE-lab/QCatch/test.yaml?branch=main" /></a>
<a href="https://COMBINE-lab.github.io/QCatch"><img alt="Documentation" src="https://img.shields.io/badge/docs-online-blue" /></a></p>
<p>QCatch: Automated quality control of single-cell quantifications from <code>alevin-fry</code> and <code>simpleaf</code>.</p>
<p>View the complete <a href="https://COMBINE-lab.github.io/QCatch">QCatch documentation</a> with interactive examples, FAQs, and detailed usage guides.</p>
<h2 id="installation">Installation</h2>
<p>You need to have Python 3.11, 3.12, or 3.13 installed on your system.</p>
<p>There are several alternative options to install QCatch:</p>
<h4 id="1-bioconda">1. Bioconda</h4>
<p>You can install using <a href="http://anaconda.org/">Conda</a>
from <a href="https://bioconda.github.io/">Bioconda</a>.</p>
<pre><code class="language-bash">conda install -c bioconda qcatch
</code></pre>
<h4 id="2-pypi">2. PyPI</h4>
<p>You can also install from <a href="https://pypi.org/project/qcatch/">PyPI</a> using <code>pip</code>:</p>
<pre><code class="language-bash">pip install qcatch
</code></pre>
<blockquote>
<p>Tips: If you run into environment issues, you can also use the provided Conda .yml file, which specifies the exact versions of all dependencies to ensure consistency.</p>
</blockquote>
<pre><code class="language-bash">conda env create -f qcatch_conda_env.yml
</code></pre>
<h2 id="basic-usage">Basic Usage</h2>
<p>Provide the path to the parent folder for quantification results, or the direct path to a .h5ad file generated by <code>alevin-fry</code> or <code>simpleaf</code>. <strong>QCatch</strong> will automatically scan the input path, assess data quality, and generate an interactive HTML report that can be viewed directly in your browser.</p>
<pre><code class="language-bash">qcatch \
    --input path/to/your/quantification/result \
    --output path/to/desired/QC/output/folder \ # if you want another folder for output
    --chemistry 10X_3p_v3
    --save_filtered_h5ad
</code></pre>
<p>For details on how to configure chemistries,
See <a href="#3--chemistry">chemistry section</a>.</p>
<h2 id="tutorial-run-qcatch-on-example-data">Tutorial: Run QCatch on Example data</h2>
<h3 id="step-1-download-dataset">Step 1 â€” Download Dataset</h3>
<pre><code class="language-bash">#!/bin/bash
set -e  # Exit immediately if a command exits with a non-zero status

echo &quot;ðŸ“¦ Downloading QCatch example dataset...&quot;

# Define where to run the tutorial (you can change this path if desired)
CWD=$(pwd)  # Current working directory
TUTORIAL_DIR=&quot;${CWD}/qcatch_tutorial&quot;

# Clean any existing tutorial directory to ensure a fresh download
rm -rf &quot;$TUTORIAL_DIR&quot; &amp;&amp; mkdir -p &quot;$TUTORIAL_DIR&quot;
ZIP_FILE=&quot;data.zip&quot;

# Download from Box
wget -O &quot;$ZIP_FILE&quot; &quot;https://umd.box.com/shared/static/zd4sai70uw9fs24e1qx6r41ec50pf45g.zip?dl=1&quot;

# Unzip and clean up
unzip &quot;$ZIP_FILE&quot; -d &quot;$TUTORIAL_DIR&quot;
rm &quot;$ZIP_FILE&quot;

echo &quot;âœ… Test data downloaded to $TUTORIAL_DIR&quot;
</code></pre>
<h3 id="step-2-run-the-qcatch">Step 2 - Run the qcatch</h3>
<p>ðŸŽ‰ All set! Now letâ€™s run QCatch:</p>
<pre><code class="language-bash">#Set up output directory
OUT_DIR=&quot;${TUTORIAL_DIR}/output&quot;
mkdir -p &quot;$OUT_DIR&quot;

# Step2 - Run QCatch
qcatch --input ${TUTORIAL_DIR}/test_data/simpleaf_with_map/quants.h5ad \
       --output ${OUT_DIR} \
</code></pre>
<h3 id="tips">Tips</h3>
<h4 id="1-input-path">1- Input path:</h4>
<p>Provide either:</p>
<ul>
<li>the <strong>path to the parent directory</strong> containing quantification results, or</li>
<li>the <strong>direct path to a .h5ad file</strong> generated by those tools.</li>
</ul>
<p>QCatch will automatically detect the input type:
- If a <strong>.h5ad file</strong> is provided, QCatch will process it directly.
- If a <strong>directory</strong> is provided, QCatch will first look for an existing .h5ad file inside. If not found, it will fall back to processing the mtx-based quantification results.</p>
<p>See the example directory structures at the end of the Tips section for reference:</p>
<h4 id="2-output-path">2- Output path:</h4>
<p>If you do not want any modifications in your input folder/files, speaficy the output path, we will save any new results and QC HTML report there.</p>
<p><strong><em>By default</em></strong>, QCatch saves the QC report and all output files in the input directory by default. If you prefer a different output location, you may specify an output path; however, this is optional.</p>
<p>Specifically:
- If QCatch detects an existing <code>quants.h5ad</code> file in the input directory and the output path is the same as the input path, QCatch will modify the original .h5ad file <em>in place</em> by appending cell-filtering results to <code>anndata.obs</code>. In addition, it will generate a separate HTML QC report in the input directory.
- For <strong>MTX-based</strong> inputs (i.e., when not using simpleaf v0.19.5 or newer), QCatch will generate a new <code>.h5ad</code> file containing metadata produced during QCatch processing. This file does <strong>NOT</strong> include metadata from the original alevin-fry quantification, which remains stored in the original files.</p>
<h4 id="3-chemistry">3- Chemistry:</h4>
<p>The <code>--chemistry</code> information is used to estimate <code>--n_partitions</code>, which represents the total partition capacity (i.e., the total number of physical droplets or wells generated in an experiment, regardless of whether they contain a cell). This value is critical for accurately defining the "ambient pool" used to model empty droplets. (<em>NOTE</em>: this is distinct from the --chemistry argument in alevin-fry, which refers to the barcode/UMI geometry.)</p>
<p>If you used a <strong>standard 10X</strong> chemistry, QCatch will first attempt to infer the chemistry from the metadata and use the internal database to get the corresponding number of partitions; If this inference fails, QCatch will stop and prompt you to explicitly provide the chemistry version using the <code>--chemistry</code> argument before rerunning the command. <strong>Supported chemistries currently include: '10X_3p_v2', '10X_3p_v3', '10X_3p_v4', '10X_3p_LT', '10X_5p_v3', or '10X_HT'</strong>.</p>
<p>For <strong>non-10x or custom</strong> assays (e.g., sci-RNA-seq3, Drop-seq), users can manually specify the capacity using <code>--n_partitions</code>. We recommend setting this value by rounding the number of processed barcodes (found in the <em>alevin-fry/simpleaf</em> log or the number of rows in the <code>.h5ad</code> file) up to the next 10% increment of the current order of magnitude. For example, If 79,000 barcodes are detected, n_partitions should be set to 80,000; If 144,000 barcodes are detected, n_partitions should be set to 150,000. This option will override any chemistry-based setting for cell-calling.</p>
<h4 id="4-gene-gene-mapping-file">4- Gene gene mapping file:</h4>
<p>If you are using simpleaf v0.19.3 or later, the generated .h5ad file already includes gene names. In this case, you do not need to specify the --gene_id2name_file option.</p>
<p>To provide a 'gene id to name mapping' info, the file should be a <strong>TSV</strong> containing two columnsâ€”â€˜gene_idâ€™ (e.g., ENSG00000284733) and â€˜gene_nameâ€™ (e.g., OR4F29)â€” <strong>without</strong> header row. If not provided, the program will attempt to retrieve the mapping from a remote registry. If that lookup fails, mitochondria plots will not be displayed, but will not affect the QC report.</p>
<h4 id="5-save-filtered-h5ad-file">5- Save filtered h5ad file:</h4>
<p>If you want to save filtered h5ad file separately, you can specify <code>--save_filtered_h5ad</code>, which is only applicable when QCatch detects the h5ad file as the input.</p>
<h4 id="6-specify-your-desired-cell-list">6- Specify your desired cell list:</h4>
<p>If you want to use a specified list of valid cell barcodes, you can provide the file path with <code>--valid_cell_list</code>. QCatch will then skip the default cell calling step and use the supplied list instead. The updated .h5ad file will include only one additional column, 'is_retained_cells', containing boolean values based on the specified list.</p>
<h4 id="7-skip-clustering-plots">7- Skip clustering plots:</h4>
<p>To reduce runtime, you may enable the <code>--skip_umap_tsne</code> option to bypass dimensionality reduction and visualization steps.</p>
<h4 id="8-export-the-summary-metrics">8- Export the summary metrics</h4>
<p>To export the summary metrics, enable the <code>--export_summary_table</code> flag. The summary table will be saved as a separate CSV file in the output directory.</p>
<h4 id="9-visualize-doublets">9- Visualize doublets</h4>
<p>To visualize doublets alongside singlets in UMAP and t-SNE plots, use both <code>--remove_doublets</code> and <code>--visualize_doublets</code> flags together. This generates two plot views:
- <strong>Retained Cells Only</strong>: Shows singlets colored by Leiden clusters (default view)
- <strong>With Doublets</strong>: Shows both singlets and doublets colored by doublet status (red for doublets, blue for singlets)</p>
<p>Toggle buttons in the HTML report allow switching between these views. Both views use the same coordinate space, so singlets appear at identical positions in both plots for easy comparison.</p>
<pre><code class="language-bash">qcatch --input path/to/quants.h5ad \
       --remove_doublets \
       --visualize_doublets
</code></pre>
<h4 id="10-debug-level-message">10- Debug-level message</h4>
<p>To get debug-level messages and more intermediate computation in cell calling step, you can specify <code>--verbose</code></p>
<h4 id="11-re-run-qcatch-on-modified-h5ad-file">11- Re-run QCatch on modified h5ad file</h4>
<p>If you re-run QCatch analysis on a modified <code>.h5ad</code> file (i.e., an <code>.h5ad</code> file with additional columns added for cell calling results), the existing cell calling-related columns will be removed and then replaced with new results. The new cell calling can be generated either through QCatch's internal method or based on a user-specified list of valid cell barcodes.</p>
<h4 id="example-directory-structures">Example directory structures:</h4>
<pre><code class="language-bash"># simpleaf
parent_quant_dir/
â”œâ”€â”€ af_map/
â”œâ”€â”€ af_quant/
â”‚   â”œâ”€â”€ alevin/
â”‚   â”‚   â”œâ”€â”€ quants_mat_cols.txt
â”‚   â”‚   â”œâ”€â”€ quants_mat_rows.txt
â”‚   â”‚   â”œâ”€â”€ quants_mat.mtx
â”‚   â”‚   â””â”€â”€ quants.h5ad (available if you use simpleaf after v0.19.3)
â”‚   â”‚   ...
â”‚   â”œâ”€â”€ featureDump.txt
â”‚   â””â”€â”€ quant.json
â””â”€â”€ simpleaf_quant_log.json

# alevin-fry
parent_quant_dir/
â”œâ”€â”€ alevin/
â”‚   â”œâ”€â”€ quants_mat_cols.txt
â”‚   â”œâ”€â”€ quants_mat_rows.txt
â”‚   â””â”€â”€ quants_mat.mtx
â”œâ”€â”€ featureDump.txt
â””â”€â”€ quant.json

</code></pre>
<p>For more advanced options and usage details, see the sections below.</p>
<h2 id="command-line-arguments">Command-Line Arguments</h2>
<table>
<thead>
<tr>
<th>Flag</th>
<th>Short</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--input</code></td>
<td><code>-i</code></td>
<td><code>str</code> (Required)</td>
<td>Path to the input directory containing the quantification output files or to the H5AD file itself.</td>
</tr>
<tr>
<td><code>--output</code></td>
<td><code>-o</code></td>
<td><code>str</code>(Required)</td>
<td>Path to the output directory.</td>
</tr>
<tr>
<td><code>--chemistry</code></td>
<td><code>-c</code></td>
<td><code>str</code>(Recommended)</td>
<td>Specifies the chemistry used in the experiment, which determines the partition range for the empty_drops step. <strong>Supported options</strong>: '10X_3p_v2', '10X_3p_v3', '10X_3p_v4', '10X_5p_v3', '10X_3p_LT', '10X_HT'. If you used a standard 10X chemistry (e.g., '10X_3p_v2', '10X_3p_v3') and performed quantification with <code>simpleaf</code>(v0.19.5 or later), QCatch will try to <strong>infer</strong> the correct chemistry from the metadata. If inference fails, QCatch will stop and prompt you to provide the chemistry explicitly via this flag.</td>
</tr>
<tr>
<td><code>--save_filtered_h5ad</code></td>
<td><code>-s</code></td>
<td><code>flag</code> (Optional)</td>
<td>If enabled, <code>qcatch</code> will save a separate <code>.h5ad</code> file containing only the final retained cells.</td>
</tr>
<tr>
<td><code>--gene_id2name_file</code></td>
<td><code>-g</code></td>
<td><code>str</code> (Optional)</td>
<td>File provides a mapping from gene IDs to gene names. The file must be a TSV containing two columnsâ€”â€˜gene_idâ€™ (e.g., ENSG00000284733) and â€˜gene_nameâ€™ (e.g., OR4F29)â€”without a header row. If not provided, the program will attempt to retrieve the mapping from a remote registry. If that lookup fails, mitochondria plots will not be displayed.</td>
</tr>
<tr>
<td><code>--valid_cell_list</code></td>
<td><code>-l</code></td>
<td><code>str</code> (Optional)</td>
<td>File provides a user-specified list of valid cell barcode. The file must be a TSV containing one column with cell barcodes without a header row. If provided, qcatch will skip the internal cell calling steps and and use the supplied list instead</td>
</tr>
<tr>
<td><code>--n_partitions</code></td>
<td><code>-n</code></td>
<td><code>int</code> (Optional)</td>
<td>Number of partitions (max number of barcodes to consider for ambient estimation). Use <code>--n_partitions</code> only when working with a custom or unsupported chemistry. When provided, this value will override the chemistry-based configuration during the cell-calling step.</td>
</tr>
<tr>
<td><code>--remove_doublets</code></td>
<td><code>-d</code></td>
<td><code>flag</code> (Optional)</td>
<td>If enabled, QCatch will perform doublet detection(use <code>Scrublet</code> tool) and remove detected doublets from the cells retained after cell calling.</td>
</tr>
<tr>
<td><code>--visualize_doublets</code></td>
<td><code>-vd</code></td>
<td><code>flag</code> (Optional)</td>
<td>If enabled (requires <code>--remove_doublets</code>), generates additional UMAP and t-SNE plots showing both singlets and doublets. Toggle buttons allow switching between "Retained Cells Only" (singlets) and "With Doublets" views.</td>
</tr>
<tr>
<td><code>--skip_umap_tsne</code></td>
<td><code>-u</code></td>
<td><code>flag</code> (Optional)</td>
<td>If provided, skips generation of UMAP and t-SNE plots.</td>
</tr>
<tr>
<td><code>--export_summary_table</code></td>
<td><code>-x</code></td>
<td><code>flag</code> (Optional)</td>
<td>If enabled, QCatch will export the summary metrics as a separate CSV file.</td>
</tr>
<tr>
<td><code>--verbose</code></td>
<td><code>-b</code></td>
<td><code>flag</code> (Optional)</td>
<td>Enable verbose logging with debug-level messages.</td>
</tr>
<tr>
<td><code>--version</code></td>
<td><code>-v</code></td>
<td><code>flag</code> (Optional)</td>
<td>Display the installed version of qcatch.</td>
</tr>
</tbody>
</table>
<!-- ## Contact

For questions and help requests, you can reach out in the [scverse discourse][].
If you found a bug, please use the [issue tracker][]. -->

<!-- ## Citation

> t.b.a

[uv]: https://github.com/astral-sh/uv
[scverse discourse]: https://discourse.scverse.org/
[issue tracker]: https://github.com/ygao61/QCatch/issues
[tests]: https://github.com/ygao61/QCatch/actions/workflows/test.yaml
[documentation]: https://QCatch.readthedocs.io
[changelog]: https://QCatch.readthedocs.io/en/latest/changelog.html
[api documentation]: https://QCatch.readthedocs.io/en/latest/api.html
[pypi]: https://pypi.org/project/QCatch -->












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": ".", "features": [], "search": "assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="assets/javascripts/bundle.79ae519e.min.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>